### PR TITLE
Add C++ Solution + Editorial to Exponentiation II (Gold)

### PR DESCRIPTION
**Pull request by ryanchou-dev**
sorry for the lack of prs. ( had a lot of homework :((( )

---
_If any of the below doesn't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which includes the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making them.

### CODE REVIEW

File: solutions/cses-1095.mdx [hunk 1]
```diff
@@ -0,0 +1,61 @@
+---
+id: cses-1712
+source: CSES
+title: Exponentiation II
+author: Ryan Chou
+---
+
+## Explanation
+
+[Fermat's Little Theorem](gold/modular) tells us that $a^{p - 1} \equiv 1 \pmod{p}$, so we can calculate $a^{b^c \pmod{1e9 + 7 - 1}} \pmod{1e9+7}$ with modular exponentiation.
```
Comment (dongliu0426) [inline solutions/cses-1095.mdx:10]:
```suggestion
[Fermat's Little Theorem](/gold/modular) tells us that $a^{p - 1} \equiv 1 \pmod{p}$, so we can calculate $a^{b^c \pmod{1e9 + 7 - 1}} \pmod{1e9+7}$ with modular exponentiation.
```
fixes link

File: solutions/cses-1095.mdx [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+	long long ans = binpow(a, pow_bc, MOD);
+
+	cout << ans << '\n';
+}
+
+int main() {
+	int t;
+	cin >> t;
+
+	while (t--) {
```
Comment (SansPapyrus683) [inline solutions/cses-1095.mdx:55]: (optinal) use for loop

File: solutions/cses-1095.mdx [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+	long long pow_bc = binpow(b, c, MOD - 1);
+	long long ans = binpow(a, pow_bc, MOD);
+
+	cout << ans << '\n';
+}
+
+int main() {
+	int t;
+	cin >> t;
+
```
Comment (SansPapyrus683) [inline solutions/cses-1095.mdx:54]:
```suggestion
```

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/cses-1095.mdx [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+	while (n > 0) {
+		if (n % 2 == 1)  // if n is odd
+			res = res * x % m;
+		x = x * x % m;
+		n /= 2;  // divide by two
+	}
+	return res;
+}
+
+void solve() {
```
Comment (SansPapyrus683) [inline solutions/cses-1095.mdx:38]: (optional) just put this in main

File: solutions/cses-1095.mdx [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+const long long MOD = 1e9 + 7;
+
+long long binpow(long long x, long long n, long long m) {
+	assert(n >= 0);
+	x %= m;
+	long long res = 1;
+	while (n > 0) {
+		if (n % 2 == 1)  // if n is odd
+			res = res * x % m;
```
Comment (SansPapyrus683) [inline solutions/cses-1095.mdx:31]: braces

File: solutions/cses-1095.mdx [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+const long long MOD = 1e9 + 7;
+
+long long binpow(long long x, long long n, long long m) {
```
Comment (SansPapyrus683) [inline solutions/cses-1095.mdx:25]: wrap this in a codesnip maybe? it's covered in the module alr

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]
sorry lol

File: solutions/cses-1095.mdx [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+int main() {
+	int t;
+	cin >> t;
+	for (int i = 0; i < t; i++) {
+		long long a;
+		long long b;
+		long long c;
+
+		cin >> a >> b >> c;
```
Comment (jessechoe10) [inline solutions/cses-1095.mdx:49]:
```suggestion
		ll a, b, c; cin >> a >> b >> c;
```

File: solutions/cses-1095.mdx [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+**Time Complexity:** $\mathcal{O}(\log P)$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
```
Comment (jessechoe10) [inline solutions/cses-1095.mdx:22]:
```suggestion
using ll = long long;
```

**Review by jessechoe10 — CHANGES_REQUESTED** [review]
Change every `long long` to `ll`.

File: solutions/cses-1095.mdx (hunk 7 above)
Comment (SansPapyrus683) [reply solutions/cses-1095.mdx:49]: no?

**Review by SansPapyrus683 — APPROVED** [review]
thanks!

File: solutions/cses-1095.mdx (hunk 7 above)
Comment (jessechoe10) [reply solutions/cses-1095.mdx:49]: lmao u could write smth that took up 5 lines of code in one line

Comment (ryanchou-dev) [reply solutions/cses-1095.mdx:49]: i wrote something like this when i coded it, but changed it during formatting

File: solutions/cses-1095.mdx [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+using ll = long long;
+const long long MOD = 1e9 + 7;
+
+//BeginCodeSnip{Binary Exponentiation}
+ll binpow(ll x, ll n, ll m) {
```
Comment (dutinmeow) [inline solutions/cses-1095.mdx:26]: Replace it with something shorter, like this
```c++
long long fast_pow(long long a, long long b, long long _MOD = 1e9 + 7) { 
    long long res = 1; 
    for (a %= _MOD; b; b >>= 1) { 
        if (b & 1) 
            res = res * a % _MOD; 
        a = a * a % _MOD; 
    } 
    return res; 
}
```

**Review by dongliu0426 — APPROVED** [review]

File: solutions/cses-1095.mdx (hunk 9 above)
Comment (ryanchou-dev) [reply solutions/cses-1095.mdx:26]: that's how they do it in the module, ill just leave it as is for consistency

File: solutions/cses-1095.mdx [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+	}
+	return res;
+}
+//EndCodeSnip
+
+int main() {
+	int t;
+	cin >> t;
+	for (int i = 0; i < t; i++) {
+		ll a;
```
[the remark this answers is not in the document]
Comment (ryanchou-dev) [reply solutions/cses-1095.mdx:45]: sans wants these in diff lines

**Review by TruaShamu — APPROVED** [review]

**Pull request merged**